infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactoryTest.java

```diff
@@ -36,12 +56,174 @@
+      throws Exception {
+    prepareNamespaceToBeFoundByName(
+        "che-default",
+        new NamespaceBuilder()
+            .withNewMetadata()
+            .withName("che-default")
+            .endMetadata()
+            .withNewStatus("Active")
+            .build());
+    namespaceFactory =
+        new KubernetesNamespaceFactory("predefined", "", "", "che-default", false, clientFactory);
+
+    List<KubernetesNamespaceMeta> availableNamespaces = namespaceFactory.list();
+    assertEquals(availableNamespaces.size(), 1);
+    KubernetesNamespaceMeta defaultNamespace = availableNamespaces.get(0);
+    assertEquals(defaultNamespace.getName(), "che-default");
+    assertEquals(defaultNamespace.getAttributes().get(DEFAULT_ATTRIBUTE), "true");
+    assertEquals(defaultNamespace.getAttributes().get(PHASE_ATTRIBUTE), "Active");
+  }
+
+  @Test
+  public void shouldReturnDefaultNamespaceWhenItDoesNotExistAndUserDefinedIsNotAllowed()
+      throws Exception {
+    prepareNamespaceToBeFoundByName("che-default", null);
+
+    namespaceFactory =
+        new KubernetesNamespaceFactory("predefined", "", "", "che-default", false, clientFactory);
+
+    List<KubernetesNamespaceMeta> availableNamespaces = namespaceFactory.list();
+    assertEquals(availableNamespaces.size(), 1);
+    KubernetesNamespaceMeta defaultNamespace = availableNamespaces.get(0);
+    assertEquals(defaultNamespace.getName(), "che-default");
+    assertEquals(defaultNamespace.getAttributes().get(DEFAULT_ATTRIBUTE), "true");
+    assertNull(
+        defaultNamespace
+            .getAttributes()
+            .get(PHASE_ATTRIBUTE)); // no phase - means such namespace does not exist
+  }
+
+  @Test(
+      expectedExceptions = InfrastructureException.class,
+      expectedExceptionsMessageRegExp =
+          "Error occurred when tried to fetch default namespace. Cause: connection refused")
+  public void shouldThrownExceptionWhenFailedToGetInfoAboutDefaultNamespace() throws Exception {
+    namespaceFactory =
+        new KubernetesNamespaceFactory("predefined", "", "", "che", false, clientFactory);
+    throwOnTryToGetNamespaceByName("che", new KubernetesClientException("connection refused"));
+
+    namespaceFactory.list();
+  }
+
+  @Test
+  public void shouldReturnListOfExistingNamespacesIfUserDefinedIsAllowed() throws Exception {
+    prepareListedNamespaces(
+        Arrays.asList(
+            createNamespace("my-for-ws", "Active"),
+            createNamespace("experimental", "Terminating")));
+
+    namespaceFactory =
+        new KubernetesNamespaceFactory("predefined", "", "", null, true, clientFactory);
+
+    List<KubernetesNamespaceMeta> availableNamespaces = namespaceFactory.list();
+    assertEquals(availableNamespaces.size(), 2);
+    KubernetesNamespaceMeta forWS = availableNamespaces.get(0);
+    assertEquals(forWS.getName(), "my-for-ws");
+    assertEquals(forWS.getAttributes().get(PHASE_ATTRIBUTE), "Active");
```

Comment:
For completeness, should we also assert that `DEFAULT_ATTRIBUTE` is also null in these PVCs?

Reply:
Added